### PR TITLE
feat: Support configuring auto-dns setting

### PR DESCRIPTION
**Pull request by liangwen12year**
reason: The user wants to ignore auto dns record (DHCP dns record) via toggling the auto-dns setting.

result: The user is able to ignore auto dns record.

### CODE REVIEW

**Comment by liangwen12year** [comment]
[citest]

File: module_utils/network_lsr/argument_validator.py [hunk 1]
```diff
@@ -2461,6 +2463,14 @@ def _ipv6_is_not_configured(connection):
                     "if you need to use initscripts.",
                 )
 
+        # initscripts does not support ip.auto_dns, so raise errors when network
+        # provider is initscripts
```
Comment (richm) [inline module_utils/network_lsr/argument_validator.py:2467]: What about `PEERDNS`? https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/networking_guide/sec-configuring_ip_networking_with_ifcg_files

or is it that we just don't want to provide any new features for the initscripts provider?

File: tests/playbooks/tests_auto_dns.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
+# SPDX-License-Identifier: BSD-3-Clause
+---
+- name: Test output device of routes
```
Comment (richm) [inline tests/playbooks/tests_auto_dns.yml:3]: `Test auto_dns settings`?

File: tests/playbooks/tests_auto_dns.yml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      include_tasks: tasks/show_interfaces.yml
+    - name: Include the task 'manage_test_interface.yml'
+      include_tasks: tasks/manage_test_interface.yml
+      vars:
+        state: present
+    - name: Include the task 'assert_device_present.yml'
+      include_tasks: tasks/assert_device_present.yml
+
+    - name: Test the route or the warning log when configuring the route with
+        or without the interface name
```
Comment (richm) [inline tests/playbooks/tests_auto_dns.yml:20]: Is this really what it's testing?

File: tests/playbooks/tests_auto_dns.yml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      block:
+        - name: Configure an interface with auto_dns enabled
```
Comment (richm) [inline tests/playbooks/tests_auto_dns.yml:22]: the test uses `auto_dns: false` - shoud the name be `Configure an interface with auto_dns disabled`?

**Comment by richm** [comment]
README.md?

File: module_utils/network_lsr/argument_validator.py (hunk 1 above)
Comment (liangwen12year) [reply module_utils/network_lsr/argument_validator.py:2467]: Yes, unless the user explicitly requested the new feature for the initscripts provider.

**Comment by liangwen12year** [comment]
[citest]

**Comment by liangwen12year** [comment]
> README.md?

I updated the Readme. Thanks.

File: library/network_connections.py [hunk 5]
```diff
@@ -1173,6 +1173,18 @@ def connection_create(self, connections, idx, connection_current=None):
...
+                if ip["ipv4_ignore_auto_dns"]:
+                    s_ip4.set_property(NM.SETTING_IP_CONFIG_IGNORE_AUTO_DNS, True)
+                else:
+                    s_ip4.set_property(NM.SETTING_IP_CONFIG_IGNORE_AUTO_DNS, False)
+
+            if ip["ipv6_ignore_auto_dns"] is not None:
+                if ip["ipv6_ignore_auto_dns"]:
+                    s_ip6.set_property(NM.SETTING_IP_CONFIG_IGNORE_AUTO_DNS, True)
+                else:
+                    s_ip6.set_property(NM.SETTING_IP_CONFIG_IGNORE_AUTO_DNS, False)
```
Comment (thom311) [inline library/network_connections.py:1186]: just:
```python
s_ip4.set_property(NM.SETTING_IP_CONFIG_IGNORE_AUTO_DNS, bool(ip["ipv4_ignore_auto_dns"]))
s_ip6.set_property(NM.SETTING_IP_CONFIG_IGNORE_AUTO_DNS, bool(ip["ipv6_ignore_auto_dns"]))
```

**Comment by liangwen12year** [comment]
[citest]

**Comment by tyll** [comment]
What is the reason to have two settings for IPv4 and IPv6? The acceptance criteria in https://issues.redhat.com/browse/NMT-507 mention only one setting "auto-dns" and Nmstate also only uses one [auto-dns setting](https://nmstate.io/devel/yaml_api.html#auto-dns) and the commit message also does not clarify the reason to distinguish between IPv4 and IPv6.

**Comment by liangwen12year** [comment]
> What is the reason to have two settings for IPv4 and IPv6? The acceptance criteria in https://issues.redhat.com/browse/NMT-507 mention only one setting "auto-dns" and Nmstate also only uses one [auto-dns setting](https://nmstate.io/devel/yaml_api.html#auto-dns) and the commit message also does not clarify the reason to distinguish between IPv4 and IPv6.

The reason that we have two settings for IPv4 and IPv6 is to be compatible with NM (`ipv4.ignore-auto-dns` and `ipv6.ignore-auto-dns`) and nmstate. In nmstate, the user needs to specify the `auto-dns` under `ipv4:` or `ipv6:` section. Also, the user can get auto DNS from DHCPv4, DHCPv6, modem etc, and they might want to ignore auto DNS on Ipv4 but not on IPv6, in this case, it is better to have two settings based on the address family.

**Comment by ffmancera** [comment]
> > What is the reason to have two settings for IPv4 and IPv6? The acceptance criteria in https://issues.redhat.com/browse/NMT-507 mention only one setting "auto-dns" and Nmstate also only uses one [auto-dns setting](https://nmstate.io/devel/yaml_api.html#auto-dns) and the commit message also does not clarify the reason to distinguish between IPv4 and IPv6.
> 
> The reason that we have two settings for IPv4 and IPv6 is to be compatible with NM (`ipv4.ignore-auto-dns` and `ipv6.ignore-auto-dns`) and nmstate. In nmstate, the user needs to specify the `auto-dns` under `ipv4:` or `ipv6:` section. Also, the user can get auto DNS from DHCPv4, DHCPv6, modem etc, and they might want to ignore auto DNS on Ipv4 but not on IPv6, in this case, it is better to have two settings based on the address family. Therefore, the acceptance criteria need to be updated.

I think the commit message should explain why there are two different properties. IMO, it is fine having them as we have them on nmcli and nmstate. It is true that on Nmstate we only have `auto-dns` but it matters if is set on `ipv4` or `ipv6` section.

**Review by ffmancera — CHANGES_REQUESTED** [review]
Please modify the commit message explaining why there are two different properties for ipv4 and ipv6.

**Comment by liangwen12year** [comment]
> Please modify the commit message explaining why there are two different properties for ipv4 and ipv6.

I modified the commit message.  Thanks.

**Comment by liangwen12year** [comment]
[citest]

**Review by ffmancera — APPROVED** [review]

**Comment by ffmancera** [comment]
Checking why python2.6 unit test failed.. not related I think

**Comment by ffmancera** [comment]
Is a tox issue @liangwen12year can you check it out? I am merging this as I do not want to block the work here.

**Pull request merged**